rust/lsp-lib/src/lsp_plugin.rs

```diff
@@ -0,0 +1,458 @@
+
+                if !contains {
+                    let config = self.config.language_config.get(&language_id).unwrap();
+
+                    let client = start_new_server(
+                        config.start_command.clone(),
+                        config.start_arguments.clone(),
+                        config.extensions.clone(),
+                        language_id,
+                    );
+
+                    match client {
+                        Ok(client) => {
+                            let client_clone = client.clone();
+                            self.language_server_clients.insert(root, client);
+                            Some(client_clone)
+                        }
+                        Err(_) => None,
+                    }
+                } else {
+                    Some(self.language_server_clients.get(&root).unwrap().clone())
+                }
+            }
+            None => {
+                let config = self.config.language_config.get(&language_id).unwrap();
+
+                if config.supports_single_file {
+                    // We check if a generic client is running. Such a client
+                    // supports single files. For example, a json client or
+                    // a Python client
+                    let contains = self.language_server_clients.contains_key("generic");
+
+                    if !contains {
+                        let client = start_new_server(
+                            config.start_command.clone(),
+                            config.start_arguments.clone(),
+                            config.extensions.clone(),
+                            language_id,
+                        );
+
+                        match client {
+                            Ok(client) => {
+                                let client_clone = client.clone();
+                                self.language_server_clients
+                                    .insert("generic".to_owned(), client);
+                                Some(client_clone)
+                            }
+                            Err(_) => None,
+                        }
+                    } else {
+                        Some(self.language_server_clients.get("generic").unwrap().clone())
+                    }
+                } else {
+                    None
+                }
+            }
+        }
+    }
+
+    /// Tries to get language for the View using the extension of the document.
+    /// Only searches for the languages supported by the Language Plugin as
+    /// defined in the config
+    fn get_language_for_view(&mut self, view: &View<ChunkCache>) -> Option<String> {
+        view.get_path()
+            .and_then(|path| path.extension())
+            .and_then(|extension| extension.to_str())
+            .and_then(|extension_str| {
+                let extension = extension_str.to_string();
+                for config in &self.config.language_config {
```

Comment:
Nit: I would do `for (lang, config) in ...` and then we don't need the `.0`/`.1`.